# Objetivo

* Prever o valor das casas utilizando um modelo de regressão linear simples

# Bibliotecas

In [19]:
import pandas as pd
import numpy as np

from sklearn import linear_model as lm

# Loading datas

In [2]:
#Abrindo o dataset
df=pd.read_csv('kc_house.csv')
df.head()

Unnamed: 0          id             date     price  bedrooms  bathrooms  \
0           0  7129300520  20141013T000000  221900.0       3.0       1.00   
1           1  6414100192  20141209T000000  538000.0       3.0       2.25   
2           2  5631500400  20150225T000000  180000.0       2.0       1.00   
3           3  2487200875  20141209T000000  604000.0       4.0       3.00   
4           4  1954400510  20150218T000000  510000.0       3.0       2.00   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1180      5650     1.0           0  ...      7        1180   
1         2570      7242     2.0           0  ...      7        2170   
2          770     10000     1.0           0  ...      6         770   
3         1960      5000     1.0           0  ...      7        1050   
4         1680      8080     1.0           0  ...      8        1680   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 22 columns]

In [25]:
df.shape

(21613, 22)

# Data Preparation

In [3]:
#features
x_train = df.drop(['price','date'], axis=1)
#Response variable
y_train = df['price'].copy()

In [4]:
x_train.head()

Unnamed: 0          id  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0           0  7129300520       3.0       1.00         1180      5650     1.0   
1           1  6414100192       3.0       2.25         2570      7242     2.0   
2           2  5631500400       2.0       1.00          770     10000     1.0   
3           3  2487200875       4.0       3.00         1960      5000     1.0   
4           4  1954400510       3.0       2.00         1680      8080     1.0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0     0          3      7        1180              0      1955   
1           0     0          3      7        2170            400      1951   
2           0     0          3      6         770              0      1933   
3           0     0          5      7        1050            910      1965   
4           0     0          3      8        1680              0      1987   

   yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0             0    98178  47.5112 -122.257           1340        5650  
1          1991    98125  47.7210 -122.319           1690        7639  
2             0    98028  47.7379 -122.233           2720        8062  
3             0    98136  47.5208 -122.393           1360        5000  
4             0    98074  47.6168 -122.045           1800        7503

In [5]:
y_train.head()

0    221900.0
1    538000.0
2    180000.0
3    604000.0
4    510000.0
Name: price, dtype: float64

# Model Training

In [7]:
#Descrição do modelo
model_lr = lm.LinearRegression()

#Treinamento do modelo
model_lr.fit(x_train,y_train)

LinearRegression()

In [8]:
#Predições
pred = model_lr.predict(x_train)

In [11]:
pred[0:5]

array([193924.79022255, 717494.17199141, 369124.73396207, 445710.40070612,
       432691.78938079])

# Performance Metrics

In [12]:
df1 = df.copy()

In [13]:
#Predição do valor da casa calculado pelo modelo 
df1['prediction'] = pred

In [ ]:
#Diferença entre o preço real e a predição do modelo
df1['error'] = df1['price'] - df1['prediction']

In [21]:
#Calculando o erro absoluto (Transforma negativos em positivos)
df1['error_abs'] = np.abs(df1['error'])

In [38]:
# Média do erro absoluto, que demonstra o quanto o preço pode estar errado
# mean absolute error
mae = np.mean(df1['error_abs'])
print('MAE = {:.2f}'.format(mae))

MAE = 125795.78


In [32]:
#Calcular o erro percentual
df1['error_perc'] = ((df1['price'] - df1['prediction']) / df1["price"])*100

In [33]:
#Calcular o erro percentual absoluto
df1['error_perc_abs'] = np.abs(df1['error_perc'])

In [40]:
#Média do erro percentual absoluto
#mean absolute percentage error
mape = np.mean(df1['error_perc_abs'])
print('MAPE = {:.1f}'.format(mape))

MAPE = 25.6


In [34]:
df1[['id', 'price', 'prediction', 'error', 'error_abs', 'error_perc', 'error_perc_abs']].head()

id     price     prediction          error      error_abs  \
0  7129300520  221900.0  193924.790223   27975.209777   27975.209777   
1  6414100192  538000.0  717494.171991 -179494.171991  179494.171991   
2  5631500400  180000.0  369124.733962 -189124.733962  189124.733962   
3  2487200875  604000.0  445710.400706  158289.599294  158289.599294   
4  1954400510  510000.0  432691.789381   77308.210619   77308.210619   

   error_perc  error_perc_abs  
0   12.607125       12.607125  
1  -33.363229       33.363229  
2 -105.069297      105.069297  
3   26.206887       26.206887  
4   15.158473       15.158473

# Conclusão

* Toda vez que o modelo faz a previsão, o erro, em média, é de 125795.78 em relação ao preço real.

* Esse valor é cerca de 25.6% do valor de venda real.

* Ou seja, toda vez que é feita uma predição, estamos sujeitos a um erro de 25.6%.

* Esse modelo pode ser melhorado para que haja um MAPE menor.